### PR TITLE
Add `Last` and `DimMap` variables.

### PR DESCRIPTION
**Pull request by ulysseB**
This PR enhance variables to
- take the last value produced in a loop nest and
- handle point-to-point communication
This was previously directly handled by instructions operands.

This changes code in sevral places
- in ir/ to support the new kinds of variables
- in codegen/ to correctly assign names to variables and alias them when
  a variable is defined as the value of another.
- in model/ to compute data dependencies
- in search_space/ to enfore constraints on mapped dimensions. Most of
  the changes is moving code from choices.exh to variable.exh

### CODE REVIEW

File: src/ir/variable.rs [hunk 1]
```diff
@@ -21,16 +20,43 @@ pub struct Variable {
...
+/// This is usefull to limit the size of the search space by removing useless decisions.
+/// For example, we don't want to store in memory the operand of a store.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq)]
+pub enum MemoryLevel {
+    /// The variable must be stored in registers and the producer and consumer must not be
+    /// separated by synchronisations.
+    RegisterNoSync,
+    /// The variable must be stored in registers.
+    Register,
+    /// The variable must be stored in registers or a local, fast, memroy.
```
Comment (nicoTolly) [inline src/ir/variable.rs]: Typo

File: src/ir/variable.rs [hunk 2]
```diff
@@ -21,16 +20,43 @@ pub struct Variable {
...
+    memory_level: MemoryLevel,
+    dimensions: VecSet<ir::DimId>,
+    def_points: VecSet<ir::StmtId>,
     use_points: VecSet<ir::StmtId>,
 }
 
+/// Indicates the slowest memory level where a variable may be stored.
+///
+/// This is usefull to limit the size of the search space by removing useless decisions.
+/// For example, we don't want to store in memory the operand of a store.
```
Comment (nicoTolly) [inline src/ir/variable.rs]: Maybe add an example of a variable that can be stored in fastMem but not in SlowMem ? I cannot think of any on the fly

File: src/ir/variable.rs [hunk 3]
```diff
@@ -21,16 +20,43 @@ pub struct Variable {
...
+    pub fn new<L>(id: VarId, def: VarDef, fun: &ir::Function<L>) -> Self {
+        let t = def.t(fun);
+        let def_points = def.def_points(fun);
+        let dimensions = def.dimensions(fun);
         Variable {
             id,
             t,
             def,
+            // TODO(ulysse): allow lowering to memory.
+            memory_level: MemoryLevel::Register,
```
Comment (nicoTolly) [inline src/ir/variable.rs]: What ? Is there a way to create a variable that can be store elsewhere than in Register ?

File: src/ir/variable.rs [hunk 4]
```diff
@@ -52,70 +78,152 @@ impl Variable {
...
-    }
-
     /// Returns the type of the variable if used on the context of `function`.
-    pub fn t<L>(self, fun: &ir::Function<L>) -> ir::Type {
-        let VarDef::Inst(inst_id) = self;
-        unwrap!(fun.inst(inst_id).t())
+    pub fn t<L>(&self, fun: &ir::Function<L>) -> ir::Type {
+        match self {
+            VarDef::Inst(inst_id) => unwrap!(fun.inst(*inst_id).t()),
+            VarDef::DimMap(value_id, ..) | VarDef::Last(value_id, ..) => {
```
Comment (nicoTolly) [inline src/ir/variable.rs]: value_id should be called var_id

File: src/ir/variable.rs [hunk 5]
```diff
@@ -52,70 +78,152 @@ impl Variable {
...
+                fun.variable(*value_id).t()
```
Comment (nicoTolly) [inline src/ir/variable.rs]: Don't understand why this is not a self-reference. Do we have the guarantee that this variable if defined as inst() ? Or am I missing something ?

File: src/codegen/variable.rs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+use ir;
+use search_space::*;
+use utils::*;
+
+/// Wraps an `ir::Variable` to expose specified decisions.
+pub struct Variable<'a> {
+    variable: &'a ir::Variable,
+    t: ir::Type,
+    instantiation_dims: HashMap<ir::DimId, usize>,
+    alias: Option<Alias>,
```
[the remark this answers is not in the document]
Comment (ulysseB) [reply src/codegen/variable.rs]: You can chain aliases: A alias with B alias with C. I added a comment to explain that on the public method.

File: src/codegen/variable.rs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            other_variable,
+            dim_mapping,
+            reverse_mapping,
+        }
+    }
+}
+
+/// Add a wrapper around variables to expose specified decisions. Orders variables with an
+/// alias after the variable they depend on.
+pub fn wrap_variables<'a>(space: &'a SearchSpace) -> Vec<Variable<'a>> {
```
[the remark this answers is not in the document]
Comment (ulysseB) [reply src/codegen/variable.rs]: Done.

File: src/codegen/name_map.rs [hunk 8]
```diff
@@ -106,7 +101,7 @@ impl<'a, 'b> NameMap<'a, 'b> {
                 mem_blocks.insert(mem_block.id(), name);
             }
         }
-
+        let variables = VariableNames::create(function, namer);
```
[the remark this answers is not in the document]
Comment (ulysseB) [reply src/codegen/name_map.rs]: I aggree. However, if you don't mind, I will do that in a later PR. I expect a good part of the code to be removed with the old operand system (Inst, Mem and Reduce should all be replaced by Variable).

File: src/ir/variable.rs (hunk 5 above)
Comment (ulysseB) [reply src/ir/variable.rs]: A variable can't depend on itself.

File: src/ir/variable.rs (hunk 3 above)
Comment (ulysseB) [reply src/ir/variable.rs]: Currently no. My goal is to unify memory blocks and variables.

File: src/ir/variable.rs (hunk 2 above)
Comment (ulysseB) [reply src/ir/variable.rs]: Done.

**Pull request merged**